File: java/client/src/main/java/com/github/fernthedev/lightchat/client/event/ServerDisconnectEvent.java
package com.github.fernthedev.lightchat.client.event;


import com.github.fernthedev.lightchat.client.Client;
import com.github.fernthedev.lightchat.core.api.event.api.Event;
import com.github.fernthedev.lightchat.core.api.event.api.HandlerList;
import com.github.fernthedev.lightchat.core.packets.IllegalConnectionPacket;
import io.netty.channel.Channel;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

/**
 * Called when client has intentionally closed connection
 * May also be called when unintentional connections are closed
 *
 * Called on {@link Client#disconnect()} after the connections are closed
 *
 */
@RequiredArgsConstructor
public class ServerDisconnectEvent extends Event  {
    private static final HandlerList handlers = new HandlerList();

    @Getter
    private final Channel channel;

    @Getter
    private final DisconnectStatus disconnectStatus;

    public ServerDisconnectEvent(Channel channel, DisconnectStatus disconnectStatus, boolean async) {
        super(async);
        this.channel = channel;
        this.disconnectStatus = disconnectStatus;
    }

    @Override
    public HandlerList getHandlers() {
        return handlers;
    }

    public static HandlerList getHandlerList() {
        return handlers;
    }

    public enum DisconnectStatus {
        /**
         *
         */
        DISCONNECTED,

        /**
         * The connection has been lost
         */
        CONNECTION_LOST,

        /**
         * When the server sends an {@link IllegalConnectionPacket}
         */
        ILLEGAL_CONNECTION,

        /**
         * Connection timed out
         */
        TIMEOUT,

        /**
         * Connection received exception
         */
        EXCEPTION,
    }
}

File: java/client/src/main/java/com/github/fernthedev/lightchat/client/event/ServerDisconnectEvent.kt
package com.github.fernthedev.lightchat.client.event

import com.github.fernthedev.lightchat.core.api.Event
import com.github.fernthedev.lightchat.core.packets.IllegalConnectionPacket
import io.netty.channel.*

/**
 * Called when client has intentionally closed connection
 * May also be called when unintentional connections are closed
 *
 * Called on [Client.disconnect] after the connections are closed
 *
 */
class ServerDisconnectEvent(
    val channel: Channel,
    val disconnectStatus: DisconnectStatus, isAsynchronous: Boolean = false
) : Event(isAsynchronous) {


    enum class DisconnectStatus {
        /**
         *
         */
        DISCONNECTED,

        /**
         * The connection has been lost
         */
        CONNECTION_LOST,

        /**
         * When the server sends an [IllegalConnectionPacket]
         */
        ILLEGAL_CONNECTION,

        /**
         * Connection timed out
         */
        TIMEOUT,

        /**
         * Connection received exception
         */
        EXCEPTION
    }

}